### PR TITLE
Avoid multi-inheritance in the cuda::features2d classes

### LINKED ISSUE #12230: cudafeatures2d cannot be wrapped for Python
##### System information (version)
- OpenCV => 4.0.0-pre (master branch)
- Operating System / Platform => Independent
- Compiler => Independent

##### Detailed description
Both classes of `cuda::ORB` and `cuda::FastFeatureDetector` utilizes multiple inheritance, thus it impossible to wrap them into Python, and perhaps other languages.

### PR DESCRIPTION
**Pull request by cv3d**
resolves #12230

```
force_builders=Custom
docker_image:Custom=ubuntu-cuda:16.04
```

### CODE REVIEW

File: modules/cudafeatures2d/include/opencv2/cudafeatures2d.hpp [hunk 1]
```diff
@@ -434,11 +434,17 @@ class CV_EXPORTS FastFeatureDetector : public cv::FastFeatureDetector, public Fe
 
         FEATURE_SIZE = 7
     };
+    enum
+    {
+        TYPE_5_8 = 0, TYPE_7_12 = 1, TYPE_9_16 = 2,
+        THRESHOLD = 10000, NONMAX_SUPPRESSION = 10001, FAST_N = 10002,
```
Comment (alalek) [inline modules/cudafeatures2d/include/opencv2/cudafeatures2d.hpp]: Code duplication should be avoided.

Does "using FastFeatureDetector::TYPE_5_8;" work here?

Comment (cv3d) [reply modules/cudafeatures2d/include/opencv2/cudafeatures2d.hpp]: I tried it, but it says `error C2602: 'TYPE_9_16' is not a member of a base class of 'cv::cuda::FastFeatureDetector'`.
We can either duplicating the enum or add `cv::` to the default value, thus become `cv::FastFeatureDetector::TYPE_9_16`. What do you think?

**Comment by cv3d** [comment]
@alalek Although the motivation is to enable binding, this PR is basically to avoid using multi-inheritance, which is implicitly tested by the fact it can compile properly, thus I believe it is a complete PR, and does not need any further clutter from binding stuff. As for the bindings generator, I am planning to introduce it for `cuda::features2d` in #12234 with its tests and binding stuff after this gets merged.

**Comment by cv3d** [comment]
@alalek I can split #12234 into 10 PRs or so (or 10 commit in same PR), each for a different module, with unit tests for each module in a separated file (or accumulated in the same file, in case of several commit & 1 PR). In that case, this PR can contain its binding code and unit tests as well. Would that make things better from your perspective?

**Comment by alalek** [comment]
As you wish. Splitting make sense for completely independent tasks.
Currently, bindings from different modules depend on some common things like changes in "modules/core/misc/python/pyopencv_core.hpp". So keeping the single PR is easy way for now.

**Pull request merged**